File: ios/SsdidWallet/Domain/MDoc/MsoParser.swift
import Foundation

/// Parser for IssuerSigned and MobileSecurityObject CBOR structures.
///
/// Decodes ISO 18013-5 IssuerSigned (containing nameSpaces + issuerAuth)
/// and the MSO payload inside a COSE_Sign1 envelope.
enum MsoParser {

    /// Parse an IssuerSigned CBOR structure.
    ///
    /// Expected layout:
    /// ```
    /// {
    ///   "nameSpaces": { <ns>: [bstr-tagged IssuerSignedItem, ...], ... },
    ///   "issuerAuth": COSE_Sign1
    /// }
    /// ```
    static func parseIssuerSigned(_ cborBytes: Data) -> IssuerSigned? {
        guard let root = CborCodec.decodeDataElement(cborBytes) as? [String: Any] else {
            return nil
        }

        guard let nameSpacesRaw = root["nameSpaces"] as? [String: Any],
              let issuerAuthRaw = root["issuerAuth"] else {
            return nil
        }

        var nameSpaces = [String: [IssuerSignedItem]]()
        for (namespace, itemsAny) in nameSpacesRaw {
            guard let itemsArray = itemsAny as? [Any] else { continue }
            var items = [IssuerSignedItem]()
            for itemAny in itemsArray {
                // Each item may be a tagged bstr containing an encoded IssuerSignedItem map,
                // or already decoded as a map.
                if let itemData = itemAny as? Data,
                   let itemMap = CborCodec.decodeDataElement(itemData) as? [String: Any] {
                    if let item = parseIssuerSignedItem(itemMap) {
                        items.append(item)
                    }
                } else if let itemMap = itemAny as? [String: Any] {
                    if let item = parseIssuerSignedItem(itemMap) {
                        items.append(item)
                    }
                }
            }
            nameSpaces[namespace] = items
        }

        // issuerAuth is a COSE_Sign1 — keep as CBOR bytes
        let issuerAuth: Data
        if let authData = issuerAuthRaw as? Data {
            issuerAuth = authData
        } else {
            issuerAuth = CborCodec.encodeDataElement(issuerAuthRaw)
        }

        return IssuerSigned(nameSpaces: nameSpaces, issuerAuth: issuerAuth)
    }

    /// Parse the MobileSecurityObject from a COSE_Sign1 issuerAuth payload.
    ///
    /// COSE_Sign1 is a CBOR array: [protected, unprotected, payload, signature].
    /// The payload (index 2) contains the MSO.
    static func parseMso(_ issuerAuthBytes: Data) -> MobileSecurityObject? {
        guard let coseArray = CborCodec.decodeDataElement(issuerAuthBytes) as? [Any],
              coseArray.count >= 3 else {
            return nil
        }

        // payload is element at index 2 — should be bstr containing MSO CBOR
        guard let payloadBytes = coseArray[2] as? Data,
              let mso = CborCodec.decodeDataElement(payloadBytes) as? [String: Any] else {
            return nil
        }

        guard let version = mso["version"] as? String,
              let digestAlgorithm = mso["digestAlgorithm"] as? String,
              let digestsRaw = mso["valueDigests"] as? [String: Any],
              let deviceKeyInfoRaw = mso["deviceKeyInfo"] as? [String: Any],
              let validityInfoRaw = mso["validityInfo"] as? [String: Any] else {
            return nil
        }

        // Parse valueDigests: { namespace: { digestId: bstr } }
        var valueDigests = [String: [Int: Data]]()
        for (namespace, nsDigestsAny) in digestsRaw {
            guard let nsDigests = nsDigestsAny as? [String: Any] else { continue }
            var digests = [Int: Data]()
            for (key, value) in nsDigests {
                guard let digestId = Int(key), let digestBytes = value as? Data else { continue }
                digests[digestId] = digestBytes
            }
            valueDigests[namespace] = digests
        }

        // Parse deviceKeyInfo
        let deviceKey = deviceKeyInfoRaw["deviceKey"] as? Data ?? Data()
        let deviceKeyInfo = DeviceKeyInfo(deviceKey: deviceKey)

        // Parse validityInfo
        guard let signed = validityInfoRaw["signed"] as? String,
              let validFrom = validityInfoRaw["validFrom"] as? String,
              let validUntil = validityInfoRaw["validUntil"] as? String else {
            return nil
        }
        let validityInfo = ValidityInfo(signed: signed, validFrom: validFrom, validUntil: validUntil)

        return MobileSecurityObject(
            version: version,
            digestAlgorithm: digestAlgorithm,
            valueDigests: valueDigests,
            deviceKeyInfo: deviceKeyInfo,
            validityInfo: validityInfo
        )
    }

    // MARK: - Private

    private static func parseIssuerSignedItem(_ map: [String: Any]) -> IssuerSignedItem? {
        guard let digestId = map["digestID"] as? Int,
              let random = map["random"] as? Data,
              let elementIdentifier = map["elementIdentifier"] as? String else {
            return nil
        }
        let elementValue = map["elementValue"] ?? ""
        return IssuerSignedItem(
            digestId: digestId,
            random: random,
            elementIdentifier: elementIdentifier,
            elementValue: elementValue
        )
    }
}

File: android/app/src/main/java/my/ssdid/wallet/domain/mdoc/MsoParser.kt
package my.ssdid.wallet.domain.mdoc

import com.upokecenter.cbor.CBORObject

object MsoParser {

    fun parseIssuerSigned(cborBytes: ByteArray): IssuerSigned {
        val cbor = CBORObject.DecodeFromBytes(cborBytes)

        val nameSpacesMap = mutableMapOf<String, List<IssuerSignedItem>>()
        val nameSpacesCbor = cbor["nameSpaces"]
            ?: throw IllegalArgumentException("Missing 'nameSpaces' in IssuerSigned CBOR")

        for (nsKey in nameSpacesCbor.keys) {
            val namespace = nsKey.AsString()
            val itemsArray = nameSpacesCbor[nsKey]
            val items = mutableListOf<IssuerSignedItem>()
            for (i in 0 until itemsArray.size()) {
                val taggedItem = itemsArray[i]
                // Handle both tag-24-wrapped and unwrapped items
                val itemCbor = if (taggedItem.isTagged && taggedItem.mostInnerTag.ToInt32Checked() == 24) {
                    CBORObject.DecodeFromBytes(taggedItem.GetByteString())
                } else {
                    taggedItem
                }
                items.add(parseIssuerSignedItem(itemCbor))
            }
            nameSpacesMap[namespace] = items
        }

        val issuerAuth = cbor["issuerAuth"]
            ?: throw IllegalArgumentException("Missing 'issuerAuth' in IssuerSigned CBOR")

        return IssuerSigned(
            nameSpaces = nameSpacesMap,
            issuerAuth = issuerAuth.EncodeToBytes()
        )
    }

    fun parseMso(issuerAuthBytes: ByteArray): MobileSecurityObject {
        val coseSign1 = CBORObject.DecodeFromBytes(issuerAuthBytes)
        val inner = if (coseSign1.isTagged) coseSign1.UntagOne() else coseSign1

        val payload = inner[2] ?: throw IllegalArgumentException("Missing payload in COSE_Sign1")
        // MSO payload may be wrapped in tag 24; unwrap if present
        val payloadBytes = if (payload.isTagged) {
            payload.UntagOne().GetByteString()
        } else {
            payload.GetByteString()
        }
        val mso = CBORObject.DecodeFromBytes(payloadBytes)

        val version = mso["version"]?.AsString()
            ?: throw IllegalArgumentException("Missing 'version' in MSO")
        val digestAlgorithm = mso["digestAlgorithm"]?.AsString()
            ?: throw IllegalArgumentException("Missing 'digestAlgorithm' in MSO")

        val valueDigests = mutableMapOf<String, Map<Int, ByteArray>>()
        val digestsCbor = mso["valueDigests"]
            ?: throw IllegalArgumentException("Missing 'valueDigests' in MSO")

        for (nsKey in digestsCbor.keys) {
            val namespace = nsKey.AsString()
            val nsDigests = mutableMapOf<Int, ByteArray>()
            val nsDigestsCbor = digestsCbor[nsKey]
            for (dKey in nsDigestsCbor.keys) {
                val digestId = dKey.AsInt32Value()
                nsDigests[digestId] = nsDigestsCbor[dKey].GetByteString()
            }
            valueDigests[namespace] = nsDigests
        }

        val deviceKeyInfoCbor = mso["deviceKeyInfo"]
            ?: throw IllegalArgumentException("Missing 'deviceKeyInfo' in MSO")
        val deviceKeyCbor = deviceKeyInfoCbor["deviceKey"]
            ?: throw IllegalArgumentException("Missing 'deviceKey' in MSO deviceKeyInfo")
        val deviceKeyInfo = DeviceKeyInfo(
            deviceKey = deviceKeyCbor.GetByteString()
        )

        val validityInfoCbor = mso["validityInfo"]
            ?: throw IllegalArgumentException("Missing 'validityInfo' in MSO")
        val validityInfo = ValidityInfo(
            signed = validityInfoCbor["signed"]?.AsString()
                ?: throw IllegalArgumentException("Missing 'signed' in validityInfo"),
            validFrom = validityInfoCbor["validFrom"]?.AsString()
                ?: throw IllegalArgumentException("Missing 'validFrom' in validityInfo"),
            validUntil = validityInfoCbor["validUntil"]?.AsString()
                ?: throw IllegalArgumentException("Missing 'validUntil' in validityInfo")
        )

        return MobileSecurityObject(
            version = version,
            digestAlgorithm = digestAlgorithm,
            valueDigests = valueDigests,
            deviceKeyInfo = deviceKeyInfo,
            validityInfo = validityInfo
        )
    }

    private fun parseIssuerSignedItem(cbor: CBORObject): IssuerSignedItem {
        return IssuerSignedItem(
            digestId = (cbor["digestID"] ?: throw IllegalArgumentException("Missing 'digestID'"))
                .AsInt32Value(),
            random = (cbor["random"] ?: throw IllegalArgumentException("Missing 'random'"))
                .GetByteString(),
            elementIdentifier = (cbor["elementIdentifier"]
                ?: throw IllegalArgumentException("Missing 'elementIdentifier'")).AsString(),
            elementValue = CborCodec.fromCborObject(
                cbor["elementValue"] ?: throw IllegalArgumentException("Missing 'elementValue'")
            )
        )
    }
}
